### ISSUE TITLE
Cannot see accessory: DNS probing gets stuck

### ISSUE THREAD

**Issue by lukasmalkmus** [brutella/hap#34]
I tried the example from the README and noticed I couldn't see the accessory in the Home App. I investigated and it turns out this code stops returning after a few iterations: https://github.com/brutella/dnssd/blob/4e37bfa22baa4db533e2e2a231db031ab00cf36d/probe.go#L174.

Also, reading from the response channel never kicks off so I assume a response is never received?

I could have opened this in the `dnssd` repo but I was only able to reproduce this with `hap`.

**Comment by lukasmalkmus** [comment]
I was able to make it work by setting a write deadline on the connection, but I don't know enough about mDNS/dnssd to evaluate if this is a good solution. Why do get writes stuck, in the first place?

**Comment by lukasmalkmus** [comment]
> Could you please check if there is anything specific about the network interface as to why writing data never finishes?

I attached the log output without my modifications so it gets stuck (I added the interface name to the `Sending probe` debug log).

<details>
  <summary>Log output</summary>
DEBUG 2023/07/20 10:38:00.974509 Joined IPv4 lo0
DEBUG 2023/07/20 10:38:00.974521 Failed joining IPv4 anpi1: setsockopt: address family not supported by protocol family
DEBUG 2023/07/20 10:38:00.974525 Failed joining IPv4 anpi0: setsockopt: address family not supported by protocol family
DEBUG 2023/07/20 10:38:00.974528 Failed joining IPv4 anpi2: setsockopt: address family not supported by protocol family
DEBUG 2023/07/20 10:38:00.974532 Failed joining IPv4 ap1: setsockopt: address family not supported by protocol family
DEBUG 2023/07/20 10:38:00.974534 Joined IPv4 en0
DEBUG 2023/07/20 10:38:00.974537 Failed joining IPv4 awdl0: setsockopt: address family not supported by protocol family
DEBUG 2023/07/20 10:38:00.974540 Failed joining IPv4 llw0: setsockopt: address family not supported by protocol family
DEBUG 2023/07/20 10:38:00.974548 Joined IPv4 utun0
DEBUG 2023/07/20 10:38:00.974552 Joined IPv4 utun1
DEBUG 2023/07/20 10:38:00.974555 Joined IPv4 utun2
DEBUG 2023/07/20 10:38:00.974607 Joined IPv4 utun3
DEBUG 2023/07/20 10:38:00.974641 Joined IPv4 utun4
DEBUG 2023/07/20 10:38:00.974682 Joined IPv4 utun5
DEBUG 2023/07/20 10:38:00.974713 Joined IPv4 utun6
DEBUG 2023/07/20 10:38:00.974752 Joined IPv4 utun7
DEBUG 2023/07/20 10:38:00.974784 Joined IPv4 utun8
DEBUG 2023/07/20 10:38:00.975038 Joined IPv6 lo0
DEBUG 2023/07/20 10:38:00.975043 Joined IPv6 anpi1
DEBUG 2023/07/20 10:38:00.975045 Joined IPv6 anpi0
DEBUG 2023/07/20 10:38:00.975048 Joined IPv6 anpi2
DEBUG 2023/07/20 10:38:00.975050 Joined IPv6 ap1
DEBUG 2023/07/20 10:38:00.975053 Joined IPv6 en0
DEBUG 2023/07/20 10:38:00.975055 Joined IPv6 awdl0
DEBUG 2023/07/20 10:38:00.975058 Joined IPv6 llw0
DEBUG 2023/07/20 10:38:00.975060 Joined IPv6 utun0
DEBUG 2023/07/20 10:38:00.975063 Joined IPv6 utun1
DEBUG 2023/07/20 10:38:00.975065 Joined IPv6 utun2
DEBUG 2023/07/20 10:38:00.975067 Joined IPv6 utun3
DEBUG 2023/07/20 10:38:00.975069 Joined IPv6 utun4
DEBUG 2023/07/20 10:38:00.975072 Joined IPv6 utun5
DEBUG 2023/07/20 10:38:00.975074 Joined IPv6 utun6
DEBUG 2023/07/20 10:38:00.975075 Joined IPv6 utun7
DEBUG 2023/07/20 10:38:00.975078 Joined IPv6 utun8
DEBUG 2023/07/20 10:38:00.975390 &{Door _hap._tcp local 72F1F945097C map[c#:4 ci:6 ff:0 id:72:F1:F9:45:09:7C md:Door pv:1.0 s#:1 sf:1 sh:XXvp7Q==] 0s 55516 [] [] map[] {0 0 <nil>}}
DEBUG 2023/07/20 10:38:00.975407 Probing for host 72F1F945097C.local. and service Door._hap._tcp.local.…
DEBUG 2023/07/20 10:38:00.975595 Joined IPv4 lo0
DEBUG 2023/07/20 10:38:00.975602 Failed joining IPv4 anpi1: setsockopt: address family not supported by protocol family
DEBUG 2023/07/20 10:38:00.975605 Failed joining IPv4 anpi0: setsockopt: address family not supported by protocol family
DEBUG 2023/07/20 10:38:00.975608 Failed joining IPv4 anpi2: setsockopt: address family not supported by protocol family
DEBUG 2023/07/20 10:38:00.975611 Failed joining IPv4 ap1: setsockopt: address family not supported by protocol family
DEBUG 2023/07/20 10:38:00.975614 Joined IPv4 en0
DEBUG 2023/07/20 10:38:00.975616 Failed joining IPv4 awdl0: setsockopt: address family not supported by protocol family
DEBUG 2023/07/20 10:38:00.975619 Failed joining IPv4 llw0: setsockopt: address family not supported by protocol family
DEBUG 2023/07/20 10:38:00.975621 Joined IPv4 utun0
DEBUG 2023/07/20 10:38:00.975624 Joined IPv4 utun1
DEBUG 2023/07/20 10:38:00.975626 Joined IPv4 utun2
DEBUG 2023/07/20 10:38:00.975628 Joined IPv4 utun3
DEBUG 2023/07/20 10:38:00.975631 Joined IPv4 utun4
DEBUG 2023/07/20 10:38:00.975632 Joined IPv4 utun5
DEBUG 2023/07/20 10:38:00.975634 Joined IPv4 utun6
DEBUG 2023/07/20 10:38:00.975636 Joined IPv4 utun7
DEBUG 2023/07/20 10:38:00.975638 Joined IPv4 utun8
DEBUG 2023/07/20 10:38:00.975796 Joined IPv6 lo0
DEBUG 2023/07/20 10:38:00.975803 Joined IPv6 anpi1
DEBUG 2023/07/20 10:38:00.975806 Joined IPv6 anpi0
DEBUG 2023/07/20 10:38:00.975808 Joined IPv6 anpi2
DEBUG 2023/07/20 10:38:00.975811 Joined IPv6 ap1
DEBUG 2023/07/20 10:38:00.975813 Joined IPv6 en0
DEBUG 2023/07/20 10:38:00.975816 Joined IPv6 awdl0
DEBUG 2023/07/20 10:38:00.975818 Joined IPv6 llw0
DEBUG 2023/07/20 10:38:00.975821 Joined IPv6 utun0
DEBUG 2023/07/20 10:38:00.975823 Joined IPv6 utun1
DEBUG 2023/07/20 10:38:00.975826 Joined IPv6 utun2
DEBUG 2023/07/20 10:38:00.975828 Joined IPv6 utun3
DEBUG 2023/07/20 10:38:00.975831 Joined IPv6 utun4
DEBUG 2023/07/20 10:38:00.975832 Joined IPv6 utun5
DEBUG 2023/07/20 10:38:00.975834 Joined IPv6 utun6
DEBUG 2023/07/20 10:38:00.975836 Joined IPv6 utun7
DEBUG 2023/07/20 10:38:00.975838 Joined IPv6 utun8
DEBUG 2023/07/20 10:38:00.975861 Probing delay 182ms
DEBUG 2023/07/20 10:38:01.159637 Draining connection
DEBUG 2023/07/20 10:38:01.160926 Sending probe lo0 ;; opcode: QUERY, status: NOERROR, id: 0
;; flags:; QUERY: 2, ANSWER: 0, AUTHORITY: 4, ADDITIONAL: 0

;; QUESTION SECTION:
;Door._hap._tcp.local.  CLASS32769       ANY
;72F1F945097C.local.    CLASS32769       ANY

;; AUTHORITY SECTION:
Door._hap._tcp.local.   120     IN      SRV     0 0 55516 72F1F945097C.local.
72F1F945097C.local.     120     IN      A       127.0.0.1
72F1F945097C.local.     120     IN      AAAA    ::1
72F1F945097C.local.     120     IN      AAAA    fe80::1

DEBUG 2023/07/20 10:38:01.161192 Sending probe anpi1 ;; opcode: QUERY, status: NOERROR, id: 0
;; flags:; QUERY: 2, ANSWER: 0, AUTHORITY: 2, ADDITIONAL: 0

;; QUESTION SECTION:
;Door._hap._tcp.local.  CLASS32769       ANY
;72F1F945097C.local.    CLASS32769       ANY

;; AUTHORITY SECTION:
Door._hap._tcp.local.   120     IN      SRV     0 0 55516 72F1F945097C.local.
72F1F945097C.local.     120     IN      AAAA    fe80::14af:e9ff:fe03:b121

DEBUG 2023/07/20 10:38:01.161254 Sending probe anpi0 ;; opcode: QUERY, status: NOERROR, id: 0
;; flags:; QUERY: 2, ANSWER: 0, AUTHORITY: 2, ADDITIONAL: 0

;; QUESTION SECTION:
;Door._hap._tcp.local.  CLASS32769       ANY
;72F1F945097C.local.    CLASS32769       ANY

;; AUTHORITY SECTION:
Door._hap._tcp.local.   120     IN      SRV     0 0 55516 72F1F945097C.local.
72F1F945097C.local.     120     IN      AAAA    fe80::14af:e9ff:fe03:b120

DEBUG 2023/07/20 10:38:01.161271 Sending probe anpi2 ;; opcode: QUERY, status: NOERROR, id: 0
;; flags:; QUERY: 2, ANSWER: 0, AUTHORITY: 2, ADDITIONAL: 0

;; QUESTION SECTION:
;Door._hap._tcp.local.  CLASS32769       ANY
;72F1F945097C.local.    CLASS32769       ANY

;; AUTHORITY SECTION:
Door._hap._tcp.local.   120     IN      SRV     0 0 55516 72F1F945097C.local.
72F1F945097C.local.     120     IN      AAAA    fe80::14af:e9ff:fe03:b122

DEBUG 2023/07/20 10:38:01.161285 Sending probe ap1 ;; opcode: QUERY, status: NOERROR, id: 0
;; flags:; QUERY: 2, ANSWER: 0, AUTHORITY: 2, ADDITIONAL: 0

;; QUESTION SECTION:
;Door._hap._tcp.local.  CLASS32769       ANY
;72F1F945097C.local.    CLASS32769       ANY

;; AUTHORITY SECTION:
Door._hap._tcp.local.   120     IN      SRV     0 0 55516 72F1F945097C.local.
72F1F945097C.local.     120     IN      AAAA    fe80::bcd0:74ff:fe2f:518
</details>

Apparently, the offending interface is `ap1`, which is used when the Mac is used as access point. Anyways, here is its `ifconfig`:

```
ap1: flags=8843<UP,BROADCAST,RUNNING,SIMPLEX,MULTICAST> mtu 1500
	options=6463<RXCSUM,TXCSUM,TSO4,TSO6,CHANNEL_IO,PARTIAL_CSUM,ZEROINVERT_CSUM>
	ether be:d0:74:2f:05:18
	inet6 fe80::bcd0:74ff:fe2f:518%ap1 prefixlen 64 scopeid 0xe
	nd6 options=201<PERFORMNUD,DAD>
	media: autoselect (<unknown type>)
	status: inactive
```

I can only imagine it is enabled because of `Content Caching` being enabled in the sharing settings (which I cannot / do not want to turn off):

![Screenshot 2023-07-20 at 10 45 53](https://github.com/brutella/hap/assets/9931588/f367449b-838d-4058-9218-2b409e884bb3)

![Screenshot 2023-07-20 at 10 45 38](https://github.com/brutella/hap/assets/9931588/7852c366-1773-46d3-adfa-c4895244c1a4)

>Setting a write deadline on the connection is definitely a good idea. Happy to accept a pull request. 😉

I'll prepare a PR over the weekend for `dnssd`, but I wonder what a sane write timeout is on those connections. 1 second? 5 seconds? When having timeouts > 1 sec, would the probe mechanism benefit from sending the probes in parallel?

**Comment by brutella** [comment]
Maybe writing to the `ap1` network interface doesn't work because your Mac doesn't have an internet connection? (See warning that Internet Sharing is currently not available.)

Nevertheless there should be a reasonable write deadline to resolve this issue. 1 second seems appropriate.

**Comment by lukasmalkmus** [comment]
@brutella I created a PR for `dnssd`: https://github.com/brutella/dnssd/pull/43.

**Comment by cweagans** [comment]
I'm having the same problem. I tested with the new dnssd changes (using `go get github.com/brutella/dnssd@4d7745af74380d9e4e53ded4a7ec240f53ba2f80`) and I'm still not seeing my device show up in the Home app. 

Here's what I'm running (only testing at this point):

```go
package main

import (
	"context"
	"log"
	"os"
	"os/signal"
	"syscall"
	"time"

	dnssdlog "github.com/brutella/dnssd/log"
	"github.com/brutella/hap"
	"github.com/brutella/hap/accessory"
)

func main() {
	dnssdlog.Debug.Enable()

	a := accessory.NewSwitch(accessory.Info{
		Name:         "SOME COOL NEW SWITCH",
		Manufacturer: "Some Cool Company",
		Model:        "Some Cool Model",
		SerialNumber: "123456789",
		Firmware:     "1.0.0",
	})

	s, err := hap.NewServer(hap.NewFsStore("./db"), a.A)
	if err != nil {
		log.Panic(err)
	}

	// Log to console when client (e.g. iOS app) changes the value of the on characteristic
	a.Switch.On.OnValueRemoteUpdate(func(on bool) {
		if on == true {
			log.Println("Client changed switch to on")
		} else {
			log.Println("Client changed switch to off")
		}
	})

	// Periodically toggle the switch's on characteristic
	go func() {
		for {
			on := !a.Switch.On.Value()
			if on == true {
				log.Println("Switch is on")
			} else {
				log.Println("Switch is off")
			}
			a.Switch.On.SetValue(on)
			time.Sleep(5 * time.Second)
		}
	}()

	c := make(chan os.Signal)
	signal.Notify(c, os.Interrupt)
	signal.Notify(c, syscall.SIGTERM)

	ctx, cancel := context.WithCancel(context.Background())
	go func() {
		<-c
		signal.Stop(c)
		cancel()
	}()

	s.ListenAndServe(ctx)
}
```

**Comment by lukasmalkmus** [comment]
@cweagans I made your example work just now:

This is the `require` section of my `go.mod`:

```
require (
	github.com/brutella/dnssd v1.2.10-0.20230725082339-4d7745af7438
	github.com/brutella/hap v0.0.26
)
```

```go
package main

import (
	"context"
	"log"
	"net/http"
	"os"
	"os/signal"
	"syscall"
	"time"

	dnssdlog "github.com/brutella/dnssd/log"
	"github.com/brutella/hap"
	"github.com/brutella/hap/accessory"
)

func main() {
	dnssdlog.Debug.Enable()

	a := accessory.NewSwitch(accessory.Info{
		Name:         "SOME COOL NEW SWITCH",
		Manufacturer: "Some Cool Company",
		Model:        "Some Cool Model",
		SerialNumber: "123456789",
		Firmware:     "1.0.0",
	})

	s, err := hap.NewServer(hap.NewFsStore("./db"), a.A)
	if err != nil {
		log.Fatal(err)
	}

	// Log to console when client (e.g. iOS app) changes the value of the on
	// characteristic.
	a.Switch.On.OnValueRemoteUpdate(func(on bool) {
		if on == true {
			log.Println("Client changed switch to on")
		} else {
			log.Println("Client changed switch to off")
		}
	})

	// Periodically toggle the switch's on characteristic
	go func() {
		for {
			on := !a.Switch.On.Value()
			if on == true {
				log.Println("Switch is on")
			} else {
				log.Println("Switch is off")
			}
			a.Switch.On.SetValue(on)
			time.Sleep(5 * time.Second)
		}
	}()

	ctx, stop := signal.NotifyContext(context.Background(),
		os.Interrupt,
		syscall.SIGTERM,
	)
	defer stop()

	if err := s.ListenAndServe(ctx); err != nil && err != http.ErrServerClosed {
		log.Fatal(err)
	}
}
```

**Comment by cweagans** [comment]
I ran the exact same code on a raspberry pi (via gokrazy) and it worked. Running it on my MacBook hasn't worked no matter how long I let it run.

I noticed in the Home app that my MacBook was showing up as a speaker to add - maybe there's a conflict there somewhere?

**Comment by brutella** [comment]
This should be fixed by now in hap v0.0.28.
